extern/storage-sealing/precommit_policy.go

```diff
@@ -72,18 +74,39 @@ func (p *BasicPreCommitPolicy) Expiration(ctx context.Context, ps ...Piece) (abi
 			continue
 		}
 
-		if end == nil || *end < p.DealInfo.DealSchedule.EndEpoch {
+		if hardLimit == nil || *hardLimit < p.DealInfo.DealSchedule.EndEpoch {
 			tmp := p.DealInfo.DealSchedule.EndEpoch
-			end = &tmp
+			hardLimit = &tmp
 		}
 	}
 
-	if end == nil {
-		tmp := epoch + p.duration
-		end = &tmp
+	exp := epoch + p.duration
+	if hardLimit != nil {
+		exp = *hardLimit
+	}
+
+	exp += miner.WPoStProvingPeriod - (exp % miner.WPoStProvingPeriod) + p.provingBoundary - 1
+
+	// add / subtract out miner.WPoStProvingPeriod to allow 24h for the PC message to land, should probably be a config
+	maxEnd := epoch + policy.GetMaxSectorExpirationExtension() - miner.WPoStProvingPeriod
+	minEnd := epoch + policy.GetMinSectorExpiration() + miner.WPoStProvingPeriod
+
+	if exp > maxEnd {
+		// we need to subtract out ceil(end - maxEnd) days
+		daysToSubtract := (exp-maxEnd)/miner.WPoStProvingPeriod + 1
+		exp -= daysToSubtract * miner.WPoStProvingPeriod
+	}
+
+	if exp < minEnd {
+		// we need to add* ceil(end - maxEnd) days
+		daysToAdd := (minEnd-exp)/miner.WPoStProvingPeriod + 1
```

Comment:
Isn't this doing `floor()+1`? Ceil would be `((minEnd-exp) + miner.WPoStProvingPeriod - 1)/miner.WPoStProvingPeriod`

Reply:
I mean, in this case with 2779/2880 probability, `ceil = floor + 1` :P

Reply:
(and in the 1/2880 case, we're off by 1 day, so no biggie)